### ISSUE TITLE
Proposed API for constant-pH simulations

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1278]
I'm discussing with @bas-rustenburg (the student heading up the constant-pH work) what setting up a constant-pH simulation should look like to the OpenMM user using the `app` layer.  We'd like your input in refining these ideas.

Currently, setting up an implicit solvent simulation with the `app` layer looks like this:
```python
pdb = PDBFile('input.pdb')
forcefield = ForceField('amber99sbildn.xml', 'amber99_obc.xml')
modeller = Modeller(pdb.topology, pdb.positions)
modeller.addHydrogens(forcefield, pH=7.0)
system = forcefield.createSystem(modeller.topology, nonbondedMethod=NoCutoff)
integrator = LangevinIntegrator(300*kelvin, 5/picoseconds, 2*femtoseconds)
simulation = Simulation(modeller.topology, system, integrator)
simulation.minimizeEnergy()
simulation.step(10000)
```
Ideally, setting up a constant-pH simulation would change this process very little.

We need to specify the following important additional pieces of information:
* the fact that we want to run a constant-pH (rather than fixed protonation state) simulation
* the pH
* which residues/components are allowed to sample multiple protonation states, perhaps making the assumption that all allowed residues/components can do so unless otherwise specified when a constant-pH simulation is selected
* which constant-pH capable forcefield (or scheme for computing protomeric populations, for small molecules) is to be used
* optionally, the user should be able to tune some parameters for the constant-pH simulation (update frequencies, probabilities of joint moves, NCMC switching trajectory lengths, etc.), but it should do the right thing (e.g. automatically calibrate reference energies) by default

Here are a few possibilities for how this could work:

### Adding simulation "updaters" in analogy to "reporters"

We could easily fit what we have in [`openmm-constph`](https://github.com/choderalab/openmm-constph) if we extend `Simulation` to have `updaters` (in analogy to `reporters`) that are periodically called:
```python
pdb = PDBFile('input.pdb')
forcefield = ForceField('amber99sbildn.xml', 'amber99_obc.xml', 'amber99sbildn-constph.xml') # add additional titratable residue definitions
modeller = Modeller(pdb.topology, pdb.positions)
modeller.addHydrogens(forcefield, pH=7.0, titratable=True) # replace titratable residues with fully protonated form
system = forcefield.createSystem(modeller.topology, nonbondedMethod=NoCutoff)
integrator = LangevinIntegrator(300*kelvin, 5/picoseconds, 2*femtoseconds)
constph = TitrationUpdater(system, forcefield, modeller.topology, pH=7.0, temperature=300*kelvin, updateInterval=500, filename='constph.out')
simulation = Simulation(modeller.topology, system, integrator)
simulation.updaters.append(constph) # similar to simulation.reporter
simulation.minimizeEnergy()
constph.calibrate() # calibrate reference energies
simulation.step(10000)
```
Describing a few changes:
*  `amber99sbildn-constph.xml` is an `ffxml` file that uses a few extensions to the `ffxml` schema (#1183) that specifies titratable variants for titratable residues (in addition to some other info described later)
* The flag `titratable` defaults to `False` for standard behavior, but when set to `True`,
```python
modeller.addHydrogens(forcefield, pH=7.0, titratable=True)
```
causes the `variants` residue slots for titratable residues to be populated with their fully protonated titratable variants [instead of just `None`](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/modeller.py#L662-L666). Users can prevent certain residues from being titratable by specifying their variants manually via the `variants` argument, as usual.
* The `pHTitrator` is a type of `SimulationUpdater` that can modify the `Context` being simulated and is called with a given update interval.  This class uses `updateParametersInContext` to carry out constant pH simulations and track statistics of the constant-pH simulation.  In principle, this could be a `reporter` since it might also be expected to output data about the protonation state change statistics.
* `ForceFIeld` will be extended to load some additional information about the titratable sites in the `ffxml` extension
* `constph.calibrate()` needs to run some short simulations on reference compounds using the same nonbonded and GB parameters as are used in `system` in order to optimize the reference energies of different titration states

Benefits:
* Minimal impact to current API
* Easy for users to specify constant pH

Drawbacks:
* We are relying on `updateParametersInContext` to carry out constant-pH simulations, which may be inefficient; possibly fine for implicit but maybe impractically slow for explicit solvent
* May require API changes if we switch to more efficient models

### Transform the system to use `Custom*Forces` and a `CustomIntegrator` for constant-pH

We could instead transform the `System` object into one that uses context parameters to allow protonation states to be rapidly switched:
```python
pdb = PDBFile('input.pdb')
forcefield = ForceField('amber99sbildn.xml', 'amber99_obc.xml', 'amber99sbildn-constph.xml') # add additional titratable residue definitions
modeller = Modeller(pdb.topology, pdb.positions)
modeller.addHydrogens(forcefield, pH=7.0, titratable=True) # replace titratable residues with fully protonated form
system = forcefield.createSystem(modeller.topology, nonbondedMethod=NoCutoff)
integrator = LangevinIntegrator(300*kelvin, 5/picoseconds, 2*femtoseconds)
# We might be able to hide some of this complexity inside of `Simulation`
constph_integrator = TitrationIntegrator(system, forcefield, temperature=300*kelvin, pH=7, ncmc_steps=50)
compound_integrator = CompoundIntegrator()
compound_integrator.addIntegrator(integrator)
integrator_index = compound_integrator.addIntegrator(constph_integrator)
constph = TitrationUpdater(system, forcefield, modeller.topology, compound_integrator, integrator_index, updateInterval=50, filename='constph.out')
simulation = Simulation(modeller.topology, system, compound_integrator)
simulation.updaters.append(constph) # would need to call `constph_integrator` to do constant pH updates and write out statistics
simulation.minimizeEnergy()
constph.calibrate() # calibrate constant-pH
simulation.step(10000)
```
* `ForceField` would create a `System` object with `NonbondedForce` terms for the titratable residues converted to `CustomNonbondedForce` terms with context parameters that select the protonation state
* Much of the complexity of creating the modified system and creating a `CompoundIntegrator` could be hidden inside of `Simulation` if some extra parameters were added to `Simulation()`. 
* The `constph` here both *updates* the protonation state *and* prints out statistics, but could be split into an `updater` and a `reporter` if desired
* `constph.calibrate()` needs to run some short simulations on reference compounds using the same nonbonded and GB parameters as are used in `system` in order to optimize the reference energies of different titration states
* Some of this might be simplified by extending `CompoundIntegrator` to be able to define cycles of `nsteps1` with `integrator1` (e.g. Langevin) followed by `nsteps2` with `integrator2` (constant pH titration).

Pros:
* Allows us to more efficiently switch protonation states with context parameters and `CustomIntegrators`

Cons:
* More complicated since it juggles multiple integrators
* Modifies system to use Custom*Force terms instead of original forces; full support for PME may still require use of `updateParametersInContext` to handle reciprocal-space terms
* Some `CustomForce` implementations may need to make duplicates of atoms in residues representing different titration states and append them to `System`, which will then not match the number of atoms in the `Topology`.  To fix this, we would need to create modified trajectory reporters that are aware of this potential difference.

### Open questions: 
* Do we still want to keep support for Amber `cpin` files for users bringing in systems through Amber `inpcrd/prmtop` file readers?

**Comment by peastman** [comment]
```XML
forcefield = ForceField('amber99sbildn.xml', 'amber99_obc.xml', 'amber99sbildn-constph.xml') # add additional titratable residue definitions
```

What residue definitions would be in `amber99sbildn-constph.xml`?  Doesn't `amber99sbildn.xml` already include the fully protonated forms?  If so, those templates will already match.

```XML
modeller.addHydrogens(forcefield, pH=7.0, titratable=True) # replace titratable residues with fully protonated form
```

Is that essentially the same as saying `pH=0`?  It means anything that can be protonated should be, right?

> We are relying on updateParametersInContext to carry out constant-pH simulations, which may be inefficient; possibly fine for implicit but maybe impractically slow for explicit solvent

That's something we should test before we start.  How expensive is it to call `updateParametersInContext()`?  We should measure it and determine whether it's going to be a problem or not.  It's also possible it could be further optimized.

> More complicated since it juggles multiple integrators

Aren't you going to be doing that either way?  You said before that each Monte Carlo step might require doing an integration.  That seems like a separate question from whether you use global or per-particle parameters to control protonation states.

**Comment by jchodera** [comment]
> What residue definitions would be in `amber99sbildn-constph.xml`? Doesn't `amber99sbildn.xml` already include the fully protonated forms? If so, those templates will already match.

`amber99sbildn.xml` already includes fully protonated forms like `ASH`, `HIP`, `LYS`, `GLH`, `TYR`, and `CYS`, but:
* According to @swails, the currently supported AMBER constant-pH forms are actually `AS4`, `GL4`, `CYS`, `TYR`, `HIP`, `LYS`, `DAP`, `DCP`, `DG`, `DT, `AP`, `CP`, `G`, and `U`.  I believe some of these (like `AS4` and `GL4`) actually include *additional* proton sites to allow syn- and anti- forms to be sampled over, since the torsion barrier is to high to allow for sampling over these in normal circumstances
* `amber99sbildn.xml` is irregular about including N-and C-terminal protonated variants (`CHIP`, `NHIP` but no `NASH`), which may just be an oversight in the preparation of this file.
* `amber99sbildn.xml` does not include uncharged variants of N- and C-terminal residues, though I am not sure if these are yet included in AMBER
* If we use include titratable small molecules as well, we would still need an additional file to specify the sites to be added.
* We also need to specify some rules for how `ForceField` or `Modeller` will translate existing residues in various individual titration states into the titratable form of the residue.  It would be best to avoid hard-coding this into `Modeller` directly, but instead allow a parameter file to encode this, along with the titratable residue form.

In addition to specifying a canonical form that `ForceField` could use to construct all possible atomic sites and a default charge state (where some sites would be uncharged dummy atoms), we need a mechanism to store additional information relevant to protonation state changes:
* The parameter sets (e.g. charges) to use for each protomeric state
* The pKas of each protomeric state
* The equilibrium solvent populations of each protomeric state
* The uncalibrated reference energies for all protomeric states
* Information about how to calibrate the reference energies, which involves the equilibrium solvent populations of each protomeric state of a corresponding reference compound.  For small molecules, this is just the compound in its different protomeric forms.  For a polymeric residue, this is a modified form of the residue itself (e.g. terminally-capped amino acid variants) in each protomeric state.

The simplest way to ensure this information is fully consistent with the reference forms of the residue describing all possible atomic sites is to simply include this extra information in additional XML tags inside a `<Residue>` definition in the `ffxml` file.  Because these extra tags would otherwise be ignored by `ForceField` or other classes due to the extensible nature of XML, this is a natural way to extend the format without breaking anything.

> Is that essentially the same as saying pH=0? It means anything that can be protonated should be, right?

Not quite.  We also have to replace `ASP` or `ASH` with `AS4`, for example, which is something that you would not do unless you were running a constant-pH simulation.  My thinking was that setting `titratable=True` would cause `Modeller` to read a `dict` of mappings that `ForceField` pulled in from any `TitratableResidue`s defined in `ffxml` files, or something similar, and use this to map any residues  in [this block](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/modeller.py#L662-L666) whose variants are not already explicitly specified.

Alternatively, we could consider how to extend the nice framework you have built for determining which hydrogens should be placed based on pH, but I would definitely want to avoid having anyone specify the counterintuitive `pH=0` for a constant pH calculation that is supposed to take place at nonzero pH, even if that is an elegant way to put on all the protons defined in `hydrogens.xml` (which we may also have to extend in some way).

> That's something we should test before we start. How expensive is it to call `updateParametersInContext()`? We should measure it and determine whether it's going to be a problem or not. It's also possible it could be further optimized.

@bas-rustenburg is working on that now.  He has just implemented an automated calibration procedure, and now we can benchmark some of the implicit solvent calculations and get to benchmarking the explicit solvent calculations soon.

It's still worth discussing whether the concept of a `SimulationUpdater` to handle protonation state changes makes sense.  It's really just a `SimulationReporter` that has the surprising behavior of also modifying the `Context` as well.

> Aren't you going to be doing that either way? You said before that each Monte Carlo step might require doing an integration. That seems like a separate question from whether you use global or per-particle parameters to control protonation states.

For explicit solvent, each Monte Carlo step does require a velocity Verlet integration of `nsteps` (which is a tunable parameter).  We could either
* (1) juggle multiple integrators with `CompositeIntegrator`
* (2) do the integration in Python, which would be very slow due to sending the forces/positions every timestep
* (3) use a `CustomIntegrator` with a tunable parameter that allows us to run either stochastic dynamics or velocity Verlet dynamics depending on the value of the parameter (which controls the collision rate)

One question: If we define `integrator1` and `integrator2` and then a `CompositeIntegrator` of the two, `integrator1.step(nsteps)` no longer works or produce anything sensible, does it?

**Comment by peastman** [comment]
It sounds like it might be best to have a completely separate force field, so you would *only* include `amber99sbildn-constph.xml`.  Trying to patch the standard force field is likely to cause problems, since in some cases there could be two different templates matching the same residue.

> If we define integrator1 and integrator2 and then a CompositeIntegrator of the two, integrator1.step(nsteps) no longer works or produce anything sensible, does it?

Currently, it would work just fine.  Here's the implementation of `step()`:

```C++
void CompoundIntegrator::step(int steps) {
    integrators[currentIntegrator]->step(steps);
}
```

So all it's doing is calling `step()` on the currently selected integrator.  I don't recommend actually doing that though.  Although it currently works, I don't guarantee it will always work.

**Comment by jchodera** [comment]
> It sounds like it might be best to have a completely separate force field, so you would only include `amber99sbildn-constph.xml`. Trying to patch the standard force field is likely to cause problems, since in some cases there could be two different templates matching the same residue.

We would need an override scheme to make this work, where a subsequent `ffxml` file could overwrite the residue definitions seen earlier.

I do like the simplicity of having a single file to use for constant-pH, but it does have drawbacks:
* It is more difficult for us to compile a single `ffxml` file that has only the parts of each forcefield we want to include, especially since creating this file in particular would involve pulling from disparate source files with completely different formats, including `.lib`, `.off`, `.prep`, leap `.rc` files, and the `cpinutils.py` code that @swails wrote.  It's possible that enough of this capability is under one hood in `ParmEd` that it could be handled entirely through extensions to `ParmEd`, but it is significantly less burden on additional forcefield development if we aren't required to put all of the components we might ever need into a *single* `ffxml` file.
* Relatedly, one of the most powerful aspects of `ffxml` files is the concept of being able to "mix and match".  We want to push this format to be even friendlier to "mix and match" so that we can combine a protein forcefield, a water forcefield, maybe nucleic acids, some lipids, and some small molecules.  Allowing the flexibility to override residue definitions (when desired) with subsequent `ffxml` definitions is actually a very powerful feature.

> Currently, it would work just fine.

That's great to hear!

> Although it currently works, I don't guarantee it will always work.

Hm.  Is it something we really don't want to guarantee?  I see there being big advantages to this continuing to be supported.

**Comment by peastman** [comment]
>  Is it something we really don't want to guarantee?

That could limit what optimizations we can do in the future.  I'm not too worried about someone selecting an integrator, then calling `step()` on that integrator directly.  That seems unlikely to cause problems.  I'm more worried about someone selecting an integrator, then calling `step()` on a *different* integrator.

**Comment by jchodera** [comment]
Oh!  In light of that, it OK if we say that calling integrator.step() on
either the currently active Integrator or the CompositeIntegrator objects
will produce the desires behavior, but that calling step() on an Integrator
that is not currently active will produce undefined behavior?

I can see the advantage in hiding the user's original Integrator inside a
CompositeIntegrator in Simulation and always making sure the original
Integrator is set as active when control is returned to the user, which
this would permit without confusing the user with a CompositeIntegrator.

**Comment by swails** [comment]
> amber99sbildn.xml does not include uncharged variants of N- and C-terminal residues, though I am not sure if these are yet included in AMBER

They are not.

> Not quite. We also have to replace ASP or ASH with AS4, for example, which is something that you would not do unless you were running a constant-pH simulation.

More restrictive: it's something you would not want to do unless you were *explicitly* titrating in the acid range.  If you were titrating at basic pHs, you'd still want to use ASP or GLU.

**Comment by peastman** [comment]
> OK if we say that calling integrator.step() on either the currently active Integrator or the CompositeIntegrator objects will produce the desires behavior

That ought to be safe.

**Comment by jchodera** [comment]
> More restrictive: it's something you would not want to do unless you were explicitly titrating in the acid range. If you were titrating at basic pHs, you'd still want to use ASP or GLU.

Can you elaborate, @swails?

**Comment by swails** [comment]
If you are simulating at a pH range of, say, 7 to 10, none of the carboxylate groups will be protonated... ever.  So you waste time titrating them.  If you just wanted to titrate the histidine, lysine, and tyrosine side-chains, you would need to keep the aspartates and glutamates as their deprotonated forms (rather than converting them to AS4 and GL4).

**Comment by jchodera** [comment]
> If you are simulating at a pH range of, say, 7 to 10, none of the carboxylate groups will be protonated... ever. So you waste time titrating them. 

We are adding constant pH capability because we don't want the user to have to worry about what is protonated and what isn't protonated.  If you want to use knowledge of the intrinsic pKas to reduce time wasted in sampling them, you use the [trick from Benoît Roux](http://pubs.acs.org/doi/abs/10.1021/acs.jctc.5b00261) where you use the intrinsic pKas to bias the selection of residues for titration and correct for this in the Metropolis-Hastings acceptance criteria.

In fact, there is good evidence to indicate that the Asp in the DFG loop of kinases like Abl [has a functionally relevant protonated form](http://www.pnas.org/content/106/1/139.full) that is critical to sampling DFG loop conformational changes.  Note that, at pH 7, the amount of free energy required to protonate an Asp existing in a "reference amino acid" form in solvent is only `ln(10) * (7 - 3.90)` ~ 7.1 kT.  This is not insurmountable, especially when you get crazy electrostatic situations like two Asps forced into close proximity to the carboxylate group of orotic acid in the [active site of orotidine decarboxylase](http://pubs.acs.org/doi/abs/10.1021/ja011096f).  One of those guys is picking up a proton.

**Comment by jchodera** [comment]
@peastman: @bas-rustenberg has done some benchmarking of the current [constant-pH implementation](https://github.com/choderalab/openmm-constph/pull/11).

We chose two benchmark systems to start with:
* DHFR in implicit solvent (154 residues, 2524 atoms)
* Abl kinase in implicit solvent (252 residues, 4200 atoms)

and will add a smaller and larger system soon.

Here are the timings for the different main parts of the code:
* abl-implicit: 0.001861 s/timestep, 0.072880 s/titration trial
* dhfr-implicit: 0.000796 s/timestep, 0.061490 s/titrationtrial

Where "timestep" denotes the effective time per call to `integrator.step(1)` and "titration" denotes the time to carry out one protonation state change proposal/acceptance cycle using the `updateParametersInContext` scheme.

For the smaller protein (DHFR), the MC titration is 77x more expensive than a single MD step; for the larger (Abl), MC titration is 39x more expensive than a single MD step.

Total time spent in each routine was quantified by profiling: 
* [benchmark notes](https://github.com/choderalab/openmm-constph/blob/benchmarks/benchmarks/notes.md)
* call graphs: [dhfr](https://raw.githubusercontent.com/choderalab/openmm-constph/benchmarks/benchmarks/dhfr-implicit/benchmark.png) [abl](https://raw.githubusercontent.com/choderalab/openmm-constph/benchmarks/benchmarks/abl-implicit/benchmark.png)

Note that we have not yet made optimizations where we strip units from the quantities and cache them to minimize the overhead from unit conversions.  That is an obvious next step.

**Comment by peastman** [comment]
In these tests, do you know what `nattempts_per_update` was set to?  If I'm reading the code correctly, each attempt involves setting various parameters, calling `updateParametersInContext()` on some number of Force objects, computing the energy, and possibly resetting all the parameters again.  And one update consists of repeating that some number of times.  So if the number of attempts is very large, it's hardly surprising that this is slow compared to one time step.

From the call graphs, it looks like the majority of the time is spent in `setTitrationState()`.  Of that time, a little under half is in OpenMM methods (`setParticleParameters()`, `setExceptionParameters()`, `updateParametersInContext()`), and the rest in Python code.  So optimizing the Python code (like by eliminating units) has the potential to roughly cut the cost of `setTitrationState()` in half.  And then we can look at the OpenMM methods and see whether anything can be done to make them faster.

**Comment by jchodera** [comment]
> In these tests, do you know what nattempts_per_update was set to?

I believe @bas-rustenburg set this to 10, but these quantities are all
per-timestep or per-update cycle, where this number was divided out.

**Comment by bas-rustenburg** [comment]
`nattempts_per_update` was (automatically/implicitly) set to 30 for dhfr, and 39 for abl, which is equal to the amount of titratable residues in each protein. As @jchodera correctly mentions these numbers were divided out. Speedup could be gained by lowering that, but at a loss of mixing of protonation states. 

As @jchodera and @peastman mention, there is room for optimization in the python code so decrease the total cost of updates. 

On the other end, I'm working on estimating the statistical efficiency of the update function and working on methods that improve it, such as improving reference energies through optimally-adjusted mixture sampling methods, weighing the choice of titratable sites by their pKa (as done in Chen & Roux 2015), and cleverer ways to select multiple residues to titrate to increase acceptance rates, to reduce the number of updates necessary.

**Comment by peastman** [comment]
Looking at the call graphs again, I notice that it only goes down as far as the Python wrapping functions.  I suspect that's where a lot of the time is being spent.  For example, it says that 17.69% of the time is spent inside `_openmm.NonbondedForce_setExceptionParameters()`.  That function does two things: convert the arguments from Python objects to C++ values, then invoke `setExceptionParameters()`.  The latter is *really* simple.  Here's its complete implementation:

```C++
    ASSERT_VALID_INDEX(index, exceptions);
    exceptions[index].particle1 = particle1;
    exceptions[index].particle2 = particle2;
    exceptions[index].chargeProd = chargeProd;
    exceptions[index].sigma = sigma;
    exceptions[index].epsilon = epsilon;
```

So I suspect it's the unwrapping of arguments that really takes most of the time.  If so, it might eventually be worth using SWIG or Cython to implement `setTitrationState()` in C++.

**Comment by jchodera** [comment]
I wonder if just stripping the units and caching the unit-stripped values would get us most of that benefit without requiring the SWIG/Cython implementation.

@bas-rustenburg: Can you give this a try?

**Comment by swails** [comment]
> I wonder if just stripping the units and caching the unit-stripped values would get us most of that benefit without requiring the SWIG/Cython implementation.

If you just make sure that all units are in the desired unit system (e.g., using `in_units_of` so that it keeps units, but changes the base unit to the one you use on a routine basis), then you can access the `._value` attribute of the `Quantity` to bypass the usual dimensional analysis.

It would certainly be worth trying since it's so simple, but I think there's a decent bit of overhead involved if you're doing this repeatedly and frequently -- at least compared to how fast OpenMM runs on GPUs.

**Comment by bas-rustenburg** [comment]
>I wonder if just stripping the units and caching the unit-stripped values would get us most of that benefit without requiring the SWIG/Cython implementation.

@jchodera I'll try it out. For future reference, it might be interesting to consider what becomes of https://github.com/pandegroup/openmm/issues/1243

**Comment by peastman** [comment]
I'm starting to look more seriously into implementing constant pH.  I'd like to pin down more clearly what the requirements are.  I'd be grateful for any thoughts on the following questions, especially from @jchodera but also anyone else interested in this feature.

1. Are we only interested in implementations that let you vary lots of sites at once (e.g. every LYS, ASP, HIS, CYS, and GLU in a large protein)?  Or would there also be value in an implementation that only varies a small number of sites chosen by the user?
2. Are we mainly interested in proteins?  Or is it also important to support varying arbitrary hydrogens on user supplied small molecules?
3. What needs to vary between states?  Charges of course, but what else?  Is it ok to keep the vdw terms on hydrogens even when they're turned off?  (This seems to be common practice.)  What about bonded terms?  Is it ok to treat them as the same in all states?  (This also seems to be common practice.)
4. How automated does it need to be?  Constant pH algorithms often require extra parameters like model pKas that are complicated to determine.  Can we say it's the user's responsibility to determine them and provide them as input?  Or do we need to be able to determine them automatically?
5. It's common to add extra ions in the solvent that get turned on or off along with sites on the solute to maintain charge neutrality.  Is it important to do that?  Or can we rely on the fact that PME automatically adds a uniform charge density to make the system neutral?

**Comment by swails** [comment]
> * Are we only interested in implementations that let you vary lots of sites at once (e.g. every LYS, ASP, HIS, CYS, and GLU in a large protein)?  Or would there also be value in an implementation that only varies a small number of sites chosen by the user?

Varying lots of sites is the most interesting (lots being, say, 10+), and varying few sites is a specific case of that.

> * Are we mainly interested in proteins?  Or is it also important to support varying arbitrary hydrogens on user supplied small molecules?

RNA molecules are of growing interest. Arbitrary hydrogens would be useful, but also quite challenging to implement in a generic way (you'd need to compute a reference energy, which in itself I would consider to be an advanced topic in its own right).

> 3\. What needs to vary between states?  Charges of course, but what else?

Lennard-Jones terms need to change. In a GB model, so too should the intrinsic radii for "inactive" protons when computing the effective Born radius.

To expand a bit here: "standard" amino acid residues (carboxylates, lysine, tyrosine, arginine, and histidine) all behave reasonably if you only change the charges between the states. However, cysteine residues behave beyond poorly due to the huge size difference between the protonated and deprotonated sulfur atoms. If you don't allow the Lennard-Jones parameters to change between states, then you cannot effectively titrate cysteine residues.

> * How automated does it need to be?  Constant pH algorithms often require extra parameters like model pKas that are complicated to determine.  Can we say it's the user's responsibility to determine them and provide them as input?  Or do we need to be able to determine them automatically?

It's reasonable to expect them to be provided (though "standard" AA residue parameters should be provided as part of the force field distribution). Reference energies for model compounds are actually pretty easy to compute - you can do a simple energy difference as an initial estimate, do a quick titration on the model compound using CpH-REMD with the energy difference as a first guess, then exactly compute (within error tolerance of your single REMD simulation) the "actual" reference energy in a single shot from the computed pKa.

There are severe complications though. For example, if protein backbone charges are allowed to change between states, then the changing 1-4 contributions across backbone atoms of neighboring residues introduces a dependency on the AA sequence on the reference compounds. Worse, if neighboring residues titrate, then the reference energy becomes both sequence- and adjacent-residue-protonation-state dependent. Even in the Amber force field where there are only 3 distinct sets of backbone charges (one for -1 residues, one for neutral residues, and one for +1 residues), the combinatorial explosion adds up. Amber worked around this by keeping the same backbone charge sets for all states of all titratable residues (specifically the "neutral residue" set, and dumping the "excess" charge on the beta-carbon). I would expect any continuous model to suffer from the same limitations as the discrete model (with which I am much more experienced).



> 5\. It's common to add extra ions in the solvent that get turned on or off along with sites on the solute to maintain charge neutrality.  Is it important to do that?  Or can we rely on the fact that PME automatically adds a uniform charge density to make the system neutral?

I've found the neutralizing plasma that PME implicitly adds to be sufficient, but I did protonation state sampling in GB, so that choice only affected conformational sampling. The main concern I'd have would be around box size. A quick test to see if it would make much impact is to run simulations of a model compound in the protonated and deprotonated form with different box sizes, and look at the resulting differences in energies between the two simulations. If you see a measurable box size dependency for "realistic" box sizes, then I would say the neutralizing plasma could introduce some artifacts and you should have a minimum recommended box size for guarding against that.

**Comment by peastman** [comment]
Thanks!  That's really helpful.  It also raises a couple of other questions.  I've been making the following assumptions.  Do they sound correct to you?

1. Ideally we want to run simulations in explicit solvent.  Using implicit solvent (either for the whole simulation or just for picking protonation states) is a hack to make the problem easier.  It's acceptable if we can't do anything better, but explicit solvent would be preferred.
2. Ideally we want to use a discrete sampling method where all your data corresponds to physically realistic states.  Continuous pH algorithms are another hack to make the problem easier.  They're acceptable if we can't do anything better, but discrete would be preferred.

**Comment by peastman** [comment]
For what it's worth, I looked to see what sorts of constant pH features are available in other MD codes.  Categorizing them by the type of algorithm (discrete or continuous) and handling of solvent (implicit, hybrid, or explicit).

| | Continuous | Discrete |
|---|---|---|
| Implicit | Amber, CHARMM | Amber, CHARMM |
| Hybrid | Amber, CHARMM | Amber |
| Explicit | Amber, CHARMM | NAMD |

There's also a continuous constant pH feature in development for Gromacs, but it hasn't made it into a release yet.

There are a couple of limitations to implementing a continuous constant pH algorithm in OpenMM.  We need to be able to compute derivatives of the energy with respect to the lambdas.  We can compute derivatives for arbitrary global parameters, but not per-particle parameters.  The way it's implemented assumes the number of global parameters is fairly small.  If you have too many of them it will run out of registers, leading to local variables getting pushed to global memory and a large performance decrease.  I haven't tested to see what point that happens at.  In any case, there could be a major performance decrease if the number of titratable sites gets too large.

The other problem is that NonbondedForce doesn't support parameter derivatives.  Adding support for them would require major new infrastructure.  You can use CustomNonbondedForce instead, but only with reaction field, not PME.

So implementing a discrete algorithm would be a lot easier.

**Comment by peastman** [comment]
I created a prototype implementation, attached here.

[constantph.py.txt](https://github.com/user-attachments/files/18327099/constantph.py.txt)

This is a very early prototype, not in any way ready for real use.  I'm posting it just to get feedback on the approach I've taken.

This is a discrete sampling, hybrid solvent approach based on @swails paper at https://pubs.acs.org/doi/10.1021/ct401042b.  It requires the following assumptions.

1. Given a starting topology, it can generate all the necessary protonation states with `Modeller.addHydrogens()`.
2. It can then generate the necessary Systems by calling `createSystem()` on ForceFields that you provide.
3. There's a one-to-one correspondence between residues and titratable sites.  Adding or removing a hydrogen can only affect parameters of atoms within that residue.  And if you have multiple hydrogens in a single residue that can vary, you need to treat them as a single site and enumerate all allowed combinations.

Here's an example script that uses it.  The constructor arguments are

- The model to simulate, specified as a Topology and atom positions.
- The ForceFields to use for creating explicit and implicit solvent systems.
- The variable residues, including the allowed variants for each one and the reference energy for each variant.
- The arguments to pass to `createSystem()` when creating explicit and implicit solvent systems.
- The integrators to use for the main simulation and for relaxing solvent after an accepted MC move.

```python
from openmm import *
from openmm.app import *
from openmm.unit import *
from constantph import ConstantPh

explicitff = ForceField('amber14-all.xml', 'amber14/tip3pfb.xml')
implicitff = ForceField('amber14-all.xml', 'implicit/obc2.xml')
pdb = PDBFile('/Users/peastman/workspace/openmm/examples/5dfr_solv-cube_equil.pdb')
residues = list(pdb.topology.residues())
variants = {
    residues[44]: ['HID', 'HIE', 'HIP'],
    residues[113]: ['HID', 'HIE', 'HIP'],
    residues[123]: ['HID', 'HIE', 'HIP'],
    residues[140]: ['HID', 'HIE', 'HIP'],
    residues[148]: ['HID', 'HIE', 'HIP'],
}
referenceEnergies = {
    residues[44]: [-11.96628184, -28.06317584, 0.0],
    residues[113]: [-11.96628184, -28.06317584, 0.0],
    residues[123]: [-11.96628184, -28.06317584, 0.0],
    residues[140]: [-11.96628184, -28.06317584, 0.0],
    residues[148]: [-11.96628184, -28.06317584, 0.0],
}
ph = ConstantPh(pdb.topology, pdb.positions, explicitff, implicitff, variants, referenceEnergies,
                {'nonbondedMethod': PME, 'nonbondedCutoff': 0.9*nanometers, 'constraints': HBonds},
                {'nonbondedMethod': CutoffNonPeriodic, 'nonbondedCutoff': 2.0*nanometers, 'constraints': HBonds},
                LangevinIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds),
                LangevinIntegrator(300.0*kelvin, 10.0/picosecond, 0.001*picosecond))
ph.simulation.minimizeEnergy()
for i in range(2000):
    ph.simulation.step(100)
    ph.attemptMCStep(300, 7.0, 500)
    print([t.currentIndex for t in ph.titrations.values()])
```

**Comment by peastman** [comment]
I have a much more complete implementation now.  I created a repository for it: https://github.com/peastman/openmm-cph.  I also wrote [a tutorial](https://github.com/peastman/openmm-cph/blob/main/Tutorial.ipynb) showing how to use it.

Anyone who is interested in constant pH simulations, I'd be very grateful if you could try it out and give feedback.  Is it a reasonable design?  How about the API?  Does it seem to be working correctly?

**Comment by peastman** [comment]
Is anybody still interested in constant pH?  It's been three weeks since I posted the first prototype, and the response is total silence.  Even if you haven't had time to try it yet, please post to say if you're still interested and plan to try it.

**Comment by ljmartin** [comment]
this is of interest to me for modeling protein-ligand interactions. looking forward to trying it out, TY.

**Comment by jchodera** [comment]
Tagging @raquellrios @sukritsingh @arianabrenner @jenkescheen @ialibay who may be interested!

**Comment by peastman** [comment]
I've found a problem I'm going to need to figure out a solution to.  It turns out that you can't run a constant pH simulation using the standard Amber ff14 force field.  It's essential that all backbone atoms have identical charges in all states of a titratable residue.  Otherwise the reference energies are sequence dependent, which means reference energies determined for a model compound can't be applied to a protein.

So Amber uses modified charges for constant pH simulations.  But they aren't stored in a standard force field file.  They're hardcoded into ParmEd, which gets called by the cpinutil.py script used to set up constant pH simulations.

I think the best solution will be to create a modified version of the ff14 XML files to be used for constant pH simulations.

**Comment by peastman** [comment]
I discussed it with @swails, and we agreed the best solution is actually to modify the script so it will automatically adjust charges as necessary.  That way it can work with any force field and any amino acids, and doesn't require the user to remember to use a different force field.

**Comment by sukritsingh** [comment]
OMG I didn't know this was here until now (thanks for tagging, John!). This is incredibly awesome and something I've been wondering about for a while, and I'm really excited to try it out. 

I'm happy to play guinea pig and test out a biological system against this protocol!

> That way it can work with any force field and any amino acids, and doesn't require the user to remember to use a different force field.

This would be an ideal scenario for me! The system I'm particularly interested in might need to have constant pH run on phosphorylated amino acids (SEP and PTR), so I would also have to use the phos-amber FF (beyond FF14), so I can see multiple force fields being needed already for a couple of systems....

Just let me know when/where to try out testing! I'll check out the plugin repo.

**Comment by peastman** [comment]
I figured out an even better solution that doesn't require changing the charges at all.  It's checked in, so everything is ready to test.

I ran some of my own tests to make sure the results are at least plausible.  Here are pKas for Hen Egg White Lysozyme, which is a standard test case in almost every paper.  It has a bunch of titratable residues, some of which have big shifts from the reference compounds.  For comparison I show the experimental values from https://onlinelibrary.wiley.com/doi/10.1002/prot.21923, and values reported for Amber in https://pubs.acs.org/doi/10.1021/ct401042b and Gromacs in https://pubs.acs.org/doi/10.1021/acs.jctc.2c00516.  There are a lot of other differences between these simulations.  That includes the type of constant pH model (discrete states with hybrid solvent for OpenMM and Amber, continuous states with explicit solvent for Gromacs), the force field (Amber14/GBn2 for OpenMM, Amber10/OBC for Amber, CHARMM36 for Gromacs), and the length of the simulations (17 ns for OpenMM, 150 ns for Amber, ~8 µs split across 105 simulations for Gromacs).  Notice that the OpenMM results are based on much less simulation time than the others, so expect the results to be a lot noisier.

|Residue|Experiment|OpenMM|Amber|Gromacs|
|---|---|---|---|---|
GLU 7|2.6-3.1|4.63|3.37|**2.82**
HIS 15|5.29-5.43|7.94|6.38|**4.84**
ASP 18|2.58-2.74|4.95|**2.83**|3.35
GLU 35|6.1-6.3|8.17|**6.27**|7.64
ASP 48|1.2-2.0|**1.26**|2.31|0.99
ASP 52|3.60-3.76|**4.28**|2.24|5.69
ASP 66|0.4-1.4|2.61|1.87|**1.7**
ASP 87|1.92-2.22|1.74|**2.02**|1.73
ASP 101|4.02-4.16|4.86|**4.36**|5.43
ASP 119|3.11-3.29|1.38|1.53|**2.77**

And here's a graph comparing them.

![Image](https://github.com/user-attachments/assets/73fcc418-27c5-4894-9f84-872d6d6b85ab)

Especially given the very short length of the OpenMM simulation, I'd say it's doing quite well.  That gives me confidence that the method is working.

**Comment by peastman** [comment]
A correction to the above: the Amber results are based on 150 ns of replica exchange with 12 replicas, so 1.8 µs total simulation time.  The OpenMM results are based on a single 17 ns run with simulated tempering, so it's about 100x less simulation time than Amber, and about 450x less simulation time than Gromacs.

**Comment by raquellrios** [comment]
This is great! 

I am currently running GROMACS constant-pH simulations with a few kinases. I would love to try it with OpenMM and compare results.

Is it already available for testing? Thanks!

**Comment by jchodera** [comment]
Those numbers all seem pretty random.

Is there a way to actually test the implementation is correct? Can you run at the pH = pKa for various individual blocked amino acids and show you can get equal populations of states? If so, that at least shows that the calibration reference free energies is appropriate for the force field and electrostatics / GB treatment combinations.

**Comment by peastman** [comment]
Oh yes, that was the very first test I did.  If I simulate the model compound, the pKa comes out right were it should.

**Comment by jchodera** [comment]
> Oh yes, that was the very first test I did. If I simulate the model compound, the pKa comes out right were it should.

Awesome! Is this a unit test, or something automated that can be run as a check? It's important to make sure that the reference free energies are still appropriate if any parameters change (e.g. force field, water model, electrostatics settings, GB model). Or, better yet, it would be useful to have a way to automatically calibrate these values for any desired combination of of these.

**Comment by peastman** [comment]
If you change any aspect of the force field, you need to recompute the reference energies.  Also if you change the temperature.

**Comment by peastman** [comment]
The implementation is now in https://github.com/openmm/openmm-cph.
